pulse/interface/viewer_3d/render_widgets/geometry_render_widget.py

```diff
@@ -191,6 +194,24 @@ def selection_callback(self, x, y):
         picked_points = self._pick_points(x, y)
         picked_structures = self._pick_structures(x, y)
 
+        for structure in picked_structures:
+            psd_parts = []
+            
+            try:
+                if "psd_name" in structure.as_dict()["extra_info"].keys():
+                    self.psd_selection_error = True
+                    psd_parts.append(structure)
+                    picked_structures = set(picked_structures)
+                    picked_structures -= set(psd_parts)
+                    picked_structures = list(picked_structures)
+                
+            except:
+                pass
```

Comment:
You can rewrite this as

```python
for structure in picked_structures:
    if "psd_name" not in structure.extra_info:
        continue
    ...
```

Reply:
Also, psd parts is always recreated in every loop.
It should be created before it.

also, the sets operation should be outside the loop.